File: CameraAccess/Managers/APIProviderManager.swift
/*
 * API Provider Manager
 * 管理不同的 API 提供商 (阿里云 Dashscope / OpenRouter)
 */

import Foundation
import SwiftUI

// MARK: - Alibaba Endpoint Enum

enum AlibabaEndpoint: String, CaseIterable, Codable {
    case beijing = "beijing"
    case singapore = "singapore"

    var displayName: String {
        switch self {
        case .beijing: return "北京 (中国大陆)"
        case .singapore: return "新加坡 (国际)"
        }
    }

    var baseURL: String {
        switch self {
        case .beijing: return "https://dashscope.aliyuncs.com/compatible-mode/v1"
        case .singapore: return "https://dashscope-intl.aliyuncs.com/compatible-mode/v1"
        }
    }

    var websocketURL: String {
        switch self {
        case .beijing: return "wss://dashscope.aliyuncs.com/api-ws/v1/realtime"
        case .singapore: return "wss://dashscope-intl.aliyuncs.com/api-ws/v1/realtime"
        }
    }
}

// MARK: - API Provider Enum (Vision API)

enum APIProvider: String, CaseIterable, Codable {
    case alibaba = "alibaba"
    case openrouter = "openrouter"

    var displayName: String {
        switch self {
        case .alibaba: return "阿里云 Dashscope"
        case .openrouter: return "OpenRouter"
        }
    }

    func baseURL(endpoint: AlibabaEndpoint = .beijing) -> String {
        switch self {
        case .alibaba: return endpoint.baseURL
        case .openrouter: return "https://openrouter.ai/api/v1"
        }
    }

    var baseURL: String {
        return baseURL(endpoint: .beijing)
    }

    var defaultModel: String {
        switch self {
        case .alibaba: return "qwen3-vl-plus"
        case .openrouter: return "google/gemini-3-flash-preview"
        }
    }

    var apiKeyHelpURL: String {
        switch self {
        case .alibaba: return "https://help.aliyun.com/zh/model-studio/get-api-key"
        case .openrouter: return "https://openrouter.ai/keys"
        }
    }

    var supportsVision: Bool {
        return true
    }
}

// MARK: - Live AI Provider Enum

enum LiveAIProvider: String, CaseIterable, Codable {
    case alibaba = "alibaba"
    case google = "google"

    var displayName: String {
        switch self {
        case .alibaba: return "阿里云 Qwen Omni"
        case .google: return "Google Gemini Live"
        }
    }

    var defaultModel: String {
        switch self {
        case .alibaba: return "qwen3-omni-flash-realtime"
        case .google: return "gemini-2.0-flash-exp"
        }
    }

    var apiKeyHelpURL: String {
        switch self {
        case .alibaba: return "https://help.aliyun.com/zh/model-studio/get-api-key"
        case .google: return "https://aistudio.google.com/apikey"
        }
    }

    func websocketURL(endpoint: AlibabaEndpoint = .beijing) -> String {
        switch self {
        case .alibaba: return endpoint.websocketURL
        case .google: return "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1beta.GenerativeService.BidiGenerateContent"
        }
    }
}

// MARK: - OpenRouter Model

struct OpenRouterModel: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let description: String?
    let contextLength: Int?
    let pricing: Pricing?
    let architecture: Architecture?

    var displayName: String {
        return name.isEmpty ? id : name
    }

    var isVisionCapable: Bool {
        // Check if model supports vision based on architecture or ID
        if let arch = architecture {
            return arch.modality?.contains("image") == true ||
                   arch.modality?.contains("multimodal") == true
        }
        // Fallback: check common vision model patterns
        let visionPatterns = ["vision", "vl", "gpt-4o", "claude-3", "gemini"]
        return visionPatterns.contains { id.lowercased().contains($0) }
    }

    var priceDisplay: String {
        guard let pricing = pricing else { return "" }
        let promptPrice = (Double(pricing.prompt) ?? 0) * 1_000_000
        let completionPrice = (Double(pricing.completion) ?? 0) * 1_000_000
        return String(format: "$%.2f / $%.2f per 1M tokens", promptPrice, completionPrice)
    }

    struct Pricing: Codable, Hashable {
        let prompt: String
        let completion: String
    }

    struct Architecture: Codable, Hashable {
        let modality: String?
        let tokenizer: String?
        let instructType: String?

        enum CodingKeys: String, CodingKey {
            case modality
            case tokenizer
            case instructType = "instruct_type"
        }
    }

    enum CodingKeys: String, CodingKey {
        case id
        case name
        case description
        case contextLength = "context_length"
        case pricing
        case architecture
    }
}

struct OpenRouterModelsResponse: Codable {
    let data: [OpenRouterModel]
}

// MARK: - API Provider Manager

@MainActor
class APIProviderManager: ObservableObject {
    static let shared = APIProviderManager()

    // Vision API Provider
    private let providerKey = "api_provider"
    private let selectedModelKey = "selected_vision_model"
    private let alibabaEndpointKey = "alibaba_endpoint"

    // Live AI Provider
    private let liveAIProviderKey = "liveai_provider"
    private let liveAIModelKey = "liveai_model"

    @Published var currentProvider: APIProvider {
        didSet {
            UserDefaults.standard.set(currentProvider.rawValue, forKey: providerKey)
            // Reset to default model when provider changes
            if oldValue != currentProvider {
                selectedModel = currentProvider.defaultModel
            }
        }
    }

    @Published var selectedModel: String {
        didSet {
            UserDefaults.standard.set(selectedModel, forKey: selectedModelKey)
        }
    }

    // Alibaba Endpoint (Beijing/Singapore)
    @Published var alibabaEndpoint: AlibabaEndpoint {
        didSet {
            UserDefaults.standard.set(alibabaEndpoint.rawValue, forKey: alibabaEndpointKey)
        }
    }

    // Live AI Provider
    @Published var liveAIProvider: LiveAIProvider {
        didSet {
            UserDefaults.standard.set(liveAIProvider.rawValue, forKey: liveAIProviderKey)
            if oldValue != liveAIProvider {
                liveAIModel = liveAIProvider.defaultModel
            }
        }
    }

    @Published var liveAIModel: String {
        didSet {
            UserDefaults.standard.set(liveAIModel, forKey: liveAIModelKey)
        }
    }

    @Published var openRouterModels: [OpenRouterModel] = []
    @Published var isLoadingModels = false
    @Published var modelsError: String?

    private init() {
        // Alibaba Endpoint
        let savedEndpoint = UserDefaults.standard.string(forKey: alibabaEndpointKey) ?? "beijing"
        self.alibabaEndpoint = AlibabaEndpoint(rawValue: savedEndpoint) ?? .beijing

        // Vision API Provider
        let savedProvider = UserDefaults.standard.string(forKey: providerKey) ?? "alibaba"
        let provider = APIProvider(rawValue: savedProvider) ?? .alibaba
        self.currentProvider = provider

        let savedModel = UserDefaults.standard.string(forKey: selectedModelKey)
        self.selectedModel = savedModel ?? provider.defaultModel

        // Live AI Provider
        let savedLiveAIProvider = UserDefaults.standard.string(forKey: liveAIProviderKey) ?? "alibaba"
        let liveProvider = LiveAIProvider(rawValue: savedLiveAIProvider) ?? .alibaba
        self.liveAIProvider = liveProvider

        let savedLiveAIModel = UserDefaults.standard.string(forKey: liveAIModelKey)
        self.liveAIModel = savedLiveAIModel ?? liveProvider.defaultModel
    }

    // MARK: - Live AI Configuration

    var liveAIWebSocketURL: String {
        return liveAIProvider.websocketURL(endpoint: alibabaEndpoint)
    }

    var liveAIAPIKey: String {
        switch liveAIProvider {
        case .alibaba:
            return APIKeyManager.shared.getAPIKey(for: .alibaba, endpoint: alibabaEndpoint) ?? ""
        case .google:
            return APIKeyManager.shared.getGoogleAPIKey() ?? ""
        }
    }

    var hasLiveAIAPIKey: Bool {
        return !liveAIAPIKey.isEmpty
    }

    // MARK: - Get Current Configuration

    var currentBaseURL: String {
        return currentProvider.baseURL(endpoint: alibabaEndpoint)
    }

    var currentAPIKey: String {
        if currentProvider == .alibaba {
            return APIKeyManager.shared.getAPIKey(for: currentProvider, endpoint: alibabaEndpoint) ?? ""
        }
        return APIKeyManager.shared.getAPIKey(for: currentProvider) ?? ""
    }

    var currentModel: String {
        return selectedModel
    }

    var hasAPIKey: Bool {
        if currentProvider == .alibaba {
            return APIKeyManager.shared.hasAPIKey(for: currentProvider, endpoint: alibabaEndpoint)
        }
        return APIKeyManager.shared.hasAPIKey(for: currentProvider)
    }

    // MARK: - OpenRouter Models

    func fetchOpenRouterModels() async {
        guard currentProvider == .openrouter else { return }
        guard let apiKey = APIKeyManager.shared.getAPIKey(for: .openrouter), !apiKey.isEmpty else {
            modelsError = "请先配置 OpenRouter API Key"
            return
        }

        isLoadingModels = true
        modelsError = nil

        do {
            let url = URL(string: "https://openrouter.ai/api/v1/models")!
            var request = URLRequest(url: url)
            request.httpMethod = "GET"
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
            request.setValue("TurboMeta", forHTTPHeaderField: "X-Title")
            request.timeoutInterval = 30

            let (data, response) = try await URLSession.shared.data(for: request)

            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                throw NSError(domain: "OpenRouter", code: -1, userInfo: [NSLocalizedDescriptionKey: "获取模型列表失败"])
            }

            let decoder = JSONDecoder()
            let modelsResponse = try decoder.decode(OpenRouterModelsResponse.self, from: data)

            // Sort models: vision-capable first, then by name
            openRouterModels = modelsResponse.data.sorted { m1, m2 in
                if m1.isVisionCapable != m2.isVisionCapable {
                    return m1.isVisionCapable
                }
                return m1.displayName < m2.displayName
            }

            print("✅ Loaded \(openRouterModels.count) OpenRouter models")

        } catch {
            modelsError = error.localizedDescription
            print("❌ Failed to fetch OpenRouter models: \(error)")
        }

        isLoadingModels = false
    }

    func searchModels(_ query: String) -> [OpenRouterModel] {
        guard !query.isEmpty else { return openRouterModels }
        let lowercaseQuery = query.lowercased()
        return openRouterModels.filter { model in
            model.id.lowercased().contains(lowercaseQuery) ||
            model.displayName.lowercased().contains(lowercaseQuery) ||
            (model.description?.lowercased().contains(lowercaseQuery) ?? false)
        }
    }

    func visionCapableModels() -> [OpenRouterModel] {
        return openRouterModels.filter { $0.isVisionCapable }
    }
}

// MARK: - Static Helpers for Non-MainActor Access

extension APIProviderManager {
    nonisolated static var staticCurrentProvider: APIProvider {
        let savedProvider = UserDefaults.standard.string(forKey: "api_provider") ?? "alibaba"
        return APIProvider(rawValue: savedProvider) ?? .alibaba
    }

    nonisolated static var staticAlibabaEndpoint: AlibabaEndpoint {
        let savedEndpoint = UserDefaults.standard.string(forKey: "alibaba_endpoint") ?? "beijing"
        return AlibabaEndpoint(rawValue: savedEndpoint) ?? .beijing
    }

    nonisolated static var staticLiveAIProvider: LiveAIProvider {
        let savedProvider = UserDefaults.standard.string(forKey: "liveai_provider") ?? "alibaba"
        return LiveAIProvider(rawValue: savedProvider) ?? .alibaba
    }

    nonisolated static var staticLiveAIAPIKey: String {
        switch staticLiveAIProvider {
        case .alibaba:
            return APIKeyManager.shared.getAPIKey(for: .alibaba, endpoint: staticAlibabaEndpoint) ?? ""
        case .google:
            return APIKeyManager.shared.getGoogleAPIKey() ?? ""
        }
    }

    nonisolated static var staticCurrentModel: String {
        let savedModel = UserDefaults.standard.string(forKey: "selected_vision_model")
        return savedModel ?? staticCurrentProvider.defaultModel
    }

    nonisolated static var staticBaseURL: String {
        return staticCurrentProvider.baseURL(endpoint: staticAlibabaEndpoint)
    }

    nonisolated static var staticAPIKey: String {
        if staticCurrentProvider == .alibaba {
            return APIKeyManager.shared.getAPIKey(for: staticCurrentProvider, endpoint: staticAlibabaEndpoint) ?? ""
        }
        return APIKeyManager.shared.getAPIKey(for: staticCurrentProvider) ?? ""
    }

    nonisolated static var staticLiveAIWebsocketURL: String {
        return staticLiveAIProvider.websocketURL(endpoint: staticAlibabaEndpoint)
    }
}

File: android/app/src/main/java/com/turbometa/rayban/managers/APIProviderManager.kt
package com.turbometa.rayban.managers

import android.content.Context
import android.content.SharedPreferences
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * API Provider Manager
 * 管理不同的 API 提供商 (阿里云 Dashscope / OpenRouter / Google)
 * 1:1 port from iOS APIProviderManager.swift
 */

// MARK: - Alibaba Endpoint Enum

enum class AlibabaEndpoint(val id: String) {
    BEIJING("beijing"),
    SINGAPORE("singapore");

    val displayName: String
        get() = when (this) {
            BEIJING -> "北京 (中国大陆)"
            SINGAPORE -> "新加坡 (国际)"
        }

    val displayNameEn: String
        get() = when (this) {
            BEIJING -> "Beijing (China)"
            SINGAPORE -> "Singapore (International)"
        }

    val baseURL: String
        get() = when (this) {
            BEIJING -> "https://dashscope.aliyuncs.com/compatible-mode/v1"
            SINGAPORE -> "https://dashscope-intl.aliyuncs.com/compatible-mode/v1"
        }

    val websocketURL: String
        get() = when (this) {
            BEIJING -> "wss://dashscope.aliyuncs.com/api-ws/v1/realtime"
            SINGAPORE -> "wss://dashscope-intl.aliyuncs.com/api-ws/v1/realtime"
        }

    companion object {
        fun fromId(id: String): AlibabaEndpoint {
            return entries.find { it.id == id } ?: BEIJING
        }
    }
}

// MARK: - API Provider Enum (Vision API)

enum class APIProvider(val id: String) {
    ALIBABA("alibaba"),
    OPENROUTER("openrouter");

    val displayName: String
        get() = when (this) {
            ALIBABA -> "阿里云 Dashscope"
            OPENROUTER -> "OpenRouter"
        }

    val displayNameEn: String
        get() = when (this) {
            ALIBABA -> "Alibaba Dashscope"
            OPENROUTER -> "OpenRouter"
        }

    fun baseURL(endpoint: AlibabaEndpoint = AlibabaEndpoint.BEIJING): String {
        return when (this) {
            ALIBABA -> endpoint.baseURL
            OPENROUTER -> "https://openrouter.ai/api/v1"
        }
    }

    val defaultModel: String
        get() = when (this) {
            ALIBABA -> "qwen-vl-flash"
            OPENROUTER -> "google/gemini-2.0-flash-001"
        }

    val apiKeyHelpURL: String
        get() = when (this) {
            ALIBABA -> "https://help.aliyun.com/zh/model-studio/get-api-key"
            OPENROUTER -> "https://openrouter.ai/keys"
        }

    val supportsVision: Boolean
        get() = true

    companion object {
        fun fromId(id: String): APIProvider {
            return entries.find { it.id == id } ?: ALIBABA
        }
    }
}

// MARK: - Live AI Provider Enum

enum class LiveAIProvider(val id: String) {
    ALIBABA("alibaba"),
    GOOGLE("google");

    val displayName: String
        get() = when (this) {
            ALIBABA -> "阿里云 Qwen Omni"
            GOOGLE -> "Google Gemini Live"
        }

    val displayNameEn: String
        get() = when (this) {
            ALIBABA -> "Alibaba Qwen Omni"
            GOOGLE -> "Google Gemini Live"
        }

    val defaultModel: String
        get() = when (this) {
            ALIBABA -> "qwen3-omni-flash-realtime"
            GOOGLE -> "gemini-2.0-flash-exp"
        }

    val apiKeyHelpURL: String
        get() = when (this) {
            ALIBABA -> "https://help.aliyun.com/zh/model-studio/get-api-key"
            GOOGLE -> "https://aistudio.google.com/apikey"
        }

    fun websocketURL(endpoint: AlibabaEndpoint = AlibabaEndpoint.BEIJING): String {
        return when (this) {
            ALIBABA -> endpoint.websocketURL
            GOOGLE -> "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1beta.GenerativeService.BidiGenerateContent"
        }
    }

    companion object {
        fun fromId(id: String): LiveAIProvider {
            return entries.find { it.id == id } ?: ALIBABA
        }
    }
}

// MARK: - OpenRouter Model

data class OpenRouterModel(
    val id: String,
    val name: String,
    val description: String? = null,
    @SerializedName("context_length")
    val contextLength: Int? = null,
    val pricing: Pricing? = null,
    val architecture: Architecture? = null
) {
    val displayName: String
        get() = name.ifEmpty { id }

    val isVisionCapable: Boolean
        get() {
            // Check if model supports vision based on architecture or ID
            architecture?.let { arch ->
                if (arch.modality?.contains("image") == true ||
                    arch.modality?.contains("multimodal") == true) {
                    return true
                }
            }
            // Fallback: check common vision model patterns
            val visionPatterns = listOf("vision", "vl", "gpt-4o", "claude-3", "gemini")
            return visionPatterns.any { id.lowercase().contains(it) }
        }

    val priceDisplay: String
        get() {
            val p = pricing ?: return ""
            val promptPrice = (p.prompt.toDoubleOrNull() ?: 0.0) * 1_000_000
            val completionPrice = (p.completion.toDoubleOrNull() ?: 0.0) * 1_000_000
            return String.format("$%.2f / $%.2f per 1M tokens", promptPrice, completionPrice)
        }

    data class Pricing(
        val prompt: String,
        val completion: String
    )

    data class Architecture(
        val modality: String? = null,
        val tokenizer: String? = null,
        @SerializedName("instruct_type")
        val instructType: String? = null
    )
}

data class OpenRouterModelsResponse(
    val data: List<OpenRouterModel>
)

// MARK: - Alibaba Vision Model

data class AlibabaVisionModel(
    val id: String,
    val displayName: String,
    val description: String
) {
    companion object {
        val availableModels = listOf(
            AlibabaVisionModel(
                "qwen-vl-flash",
                "Qwen VL Flash",
                "快速响应，适合实时场景"
            ),
            AlibabaVisionModel(
                "qwen-vl-plus",
                "Qwen VL Plus",
                "均衡性能，推荐日常使用"
            ),
            AlibabaVisionModel(
                "qwen-vl-max",
                "Qwen VL Max",
                "最强性能，适合复杂任务"
            ),
            AlibabaVisionModel(
                "qwen2.5-vl-72b-instruct",
                "Qwen 2.5 VL 72B",
                "大参数模型，高精度分析"
            )
        )
    }
}

// MARK: - API Provider Manager

class APIProviderManager private constructor(context: Context) {

    companion object {
        private const val PREFS_NAME = "api_provider_prefs"
        private const val KEY_PROVIDER = "api_provider"
        private const val KEY_SELECTED_MODEL = "selected_vision_model"
        private const val KEY_ALIBABA_ENDPOINT = "alibaba_endpoint"
        private const val KEY_LIVE_AI_PROVIDER = "liveai_provider"
        private const val KEY_LIVE_AI_MODEL = "liveai_model"

        @Volatile
        private var instance: APIProviderManager? = null

        fun getInstance(context: Context): APIProviderManager {
            return instance ?: synchronized(this) {
                instance ?: APIProviderManager(context.applicationContext).also { instance = it }
            }
        }

        // Static accessors for non-context access
        private var prefs: SharedPreferences? = null

        val staticCurrentProvider: APIProvider
            get() {
                val id = prefs?.getString(KEY_PROVIDER, "alibaba") ?: "alibaba"
                return APIProvider.fromId(id)
            }

        val staticAlibabaEndpoint: AlibabaEndpoint
            get() {
                val id = prefs?.getString(KEY_ALIBABA_ENDPOINT, "beijing") ?: "beijing"
                return AlibabaEndpoint.fromId(id)
            }

        val staticLiveAIProvider: LiveAIProvider
            get() {
                val id = prefs?.getString(KEY_LIVE_AI_PROVIDER, "alibaba") ?: "alibaba"
                return LiveAIProvider.fromId(id)
            }

        val staticCurrentModel: String
            get() {
                return prefs?.getString(KEY_SELECTED_MODEL, null)
                    ?: staticCurrentProvider.defaultModel
            }

        val staticBaseURL: String
            get() = staticCurrentProvider.baseURL(staticAlibabaEndpoint)
    }

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val gson = Gson()
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    init {
        // Set static prefs for non-context access
        Companion.prefs = this.prefs
    }

    // Vision API Provider
    private val _currentProvider = MutableStateFlow(
        APIProvider.fromId(prefs.getString(KEY_PROVIDER, "alibaba") ?: "alibaba")
    )
    val currentProvider: StateFlow<APIProvider> = _currentProvider

    private val _selectedModel = MutableStateFlow(
        prefs.getString(KEY_SELECTED_MODEL, null) ?: APIProvider.ALIBABA.defaultModel
    )
    val selectedModel: StateFlow<String> = _selectedModel

    // Alibaba Endpoint (Beijing/Singapore)
    private val _alibabaEndpoint = MutableStateFlow(
        AlibabaEndpoint.fromId(prefs.getString(KEY_ALIBABA_ENDPOINT, "beijing") ?: "beijing")
    )
    val alibabaEndpoint: StateFlow<AlibabaEndpoint> = _alibabaEndpoint

    // Live AI Provider
    private val _liveAIProvider = MutableStateFlow(
        LiveAIProvider.fromId(prefs.getString(KEY_LIVE_AI_PROVIDER, "alibaba") ?: "alibaba")
    )
    val liveAIProvider: StateFlow<LiveAIProvider> = _liveAIProvider

    private val _liveAIModel = MutableStateFlow(
        prefs.getString(KEY_LIVE_AI_MODEL, null) ?: LiveAIProvider.ALIBABA.defaultModel
    )
    val liveAIModel: StateFlow<String> = _liveAIModel

    // OpenRouter Models
    private val _openRouterModels = MutableStateFlow<List<OpenRouterModel>>(emptyList())
    val openRouterModels: StateFlow<List<OpenRouterModel>> = _openRouterModels

    private val _isLoadingModels = MutableStateFlow(false)
    val isLoadingModels: StateFlow<Boolean> = _isLoadingModels

    private val _modelsError = MutableStateFlow<String?>(null)
    val modelsError: StateFlow<String?> = _modelsError

    // MARK: - Setters

    fun setCurrentProvider(provider: APIProvider) {
        val oldValue = _currentProvider.value
        _currentProvider.value = provider
        prefs.edit().putString(KEY_PROVIDER, provider.id).apply()

        // Reset to default model when provider changes
        if (oldValue != provider) {
            setSelectedModel(provider.defaultModel)
        }
    }

    fun setSelectedModel(model: String) {
        _selectedModel.value = model
        prefs.edit().putString(KEY_SELECTED_MODEL, model).apply()
    }

    fun setAlibabaEndpoint(endpoint: AlibabaEndpoint) {
        _alibabaEndpoint.value = endpoint
        prefs.edit().putString(KEY_ALIBABA_ENDPOINT, endpoint.id).apply()
    }

    fun setLiveAIProvider(provider: LiveAIProvider) {
        val oldValue = _liveAIProvider.value
        _liveAIProvider.value = provider
        prefs.edit().putString(KEY_LIVE_AI_PROVIDER, provider.id).apply()

        if (oldValue != provider) {
            setLiveAIModel(provider.defaultModel)
        }
    }

    fun setLiveAIModel(model: String) {
        _liveAIModel.value = model
        prefs.edit().putString(KEY_LIVE_AI_MODEL, model).apply()
    }

    // MARK: - Live AI Configuration

    val liveAIWebSocketURL: String
        get() = _liveAIProvider.value.websocketURL(_alibabaEndpoint.value)

    fun getLiveAIAPIKey(apiKeyManager: com.turbometa.rayban.utils.APIKeyManager): String {
        return when (_liveAIProvider.value) {
            LiveAIProvider.ALIBABA -> apiKeyManager.getAPIKey(APIProvider.ALIBABA, _alibabaEndpoint.value) ?: ""
            LiveAIProvider.GOOGLE -> apiKeyManager.getGoogleAPIKey() ?: ""
        }
    }

    fun hasLiveAIAPIKey(apiKeyManager: com.turbometa.rayban.utils.APIKeyManager): Boolean {
        return getLiveAIAPIKey(apiKeyManager).isNotEmpty()
    }

    // MARK: - Get Current Configuration

    val currentBaseURL: String
        get() = _currentProvider.value.baseURL(_alibabaEndpoint.value)

    fun getCurrentAPIKey(apiKeyManager: com.turbometa.rayban.utils.APIKeyManager): String {
        return if (_currentProvider.value == APIProvider.ALIBABA) {
            apiKeyManager.getAPIKey(_currentProvider.value, _alibabaEndpoint.value) ?: ""
        } else {
            apiKeyManager.getAPIKey(_currentProvider.value) ?: ""
        }
    }

    val currentModel: String
        get() = _selectedModel.value

    fun hasAPIKey(apiKeyManager: com.turbometa.rayban.utils.APIKeyManager): Boolean {
        return if (_currentProvider.value == APIProvider.ALIBABA) {
            apiKeyManager.hasAPIKey(_currentProvider.value, _alibabaEndpoint.value)
        } else {
            apiKeyManager.hasAPIKey(_currentProvider.value)
        }
    }

    // MARK: - OpenRouter Models

    suspend fun fetchOpenRouterModels(apiKeyManager: com.turbometa.rayban.utils.APIKeyManager) {
        if (_currentProvider.value != APIProvider.OPENROUTER) return

        val apiKey = apiKeyManager.getAPIKey(APIProvider.OPENROUTER)
        if (apiKey.isNullOrEmpty()) {
            _modelsError.value = "请先配置 OpenRouter API Key"
            return
        }

        _isLoadingModels.value = true
        _modelsError.value = null

        withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("https://openrouter.ai/api/v1/models")
                    .addHeader("Authorization", "Bearer $apiKey")
                    .addHeader("X-Title", "TurboMeta")
                    .get()
                    .build()

                val response = httpClient.newCall(request).execute()

                if (!response.isSuccessful) {
                    _modelsError.value = "获取模型列表失败: ${response.code}"
                    return@withContext
                }

                val body = response.body?.string() ?: ""
                val modelsResponse = gson.fromJson(body, OpenRouterModelsResponse::class.java)

                // Sort models: vision-capable first, then by name
                _openRouterModels.value = modelsResponse.data.sortedWith(
                    compareByDescending<OpenRouterModel> { it.isVisionCapable }
                        .thenBy { it.displayName }
                )

                android.util.Log.d("APIProviderManager", "Loaded ${_openRouterModels.value.size} OpenRouter models")

            } catch (e: Exception) {
                _modelsError.value = e.message ?: "Unknown error"
                android.util.Log.e("APIProviderManager", "Failed to fetch OpenRouter models", e)
            }
        }

        _isLoadingModels.value = false
    }

    fun searchModels(query: String): List<OpenRouterModel> {
        if (query.isEmpty()) return _openRouterModels.value
        val lowercaseQuery = query.lowercase()
        return _openRouterModels.value.filter { model ->
            model.id.lowercase().contains(lowercaseQuery) ||
            model.displayName.lowercase().contains(lowercaseQuery) ||
            (model.description?.lowercase()?.contains(lowercaseQuery) == true)
        }
    }

    fun visionCapableModels(): List<OpenRouterModel> {
        return _openRouterModels.value.filter { it.isVisionCapable }
    }
}
